### PR TITLE
96 optimization   plsan free local variable calls internalalloc that makes plsan slow down

### PR DESCRIPTION
**Pull request by purplepig4657**
#96 

__plsan_free_local_variable calls InternalAlloc function per pointer variable in function.
InternalAlloc is dynamic allocation and too many InternalAlloc called when function return. It affects to PLSan slow down.

This pr resolve this problem with allocating one vector per thread to store all address that need to be stored in __plsan_free_local_variable. This method reduce InternalAlloc to just one time per thread.

And, there are memory leak that lazy_check didn't free all InternalAlloc memory block when llvm.stacksave and llvm.stackrestore are in another basic block. The method described above can fix this problem also.

### CODE REVIEW

File: compiler-rt/lib/plsan/plsan.cpp [hunk 1]
```diff
@@ -30,46 +30,31 @@ extern "C" void __plsan_store(void **lhs, void *rhs) {
...
-    return lazy_check_info;
-  }
 }
 
-extern "C" void __plsan_lazy_check(LazyCheckInfo *lazy_check_info,
-                                   void *ret_addr) {
+extern "C" void __plsan_lazy_check(void *ret_addr) {
   __sanitizer::Vector<void *> *lazy_check_addr_list =
-      lazy_check_info->RefCountZeroAddrs;
+      __plsan::local_var_ref_count_zero_list;
```
Comment (hygoni) [inline compiler-rt/lib/plsan/plsan.cpp:48]: I'm pretty sure that this commit will not work correctly.

How local_var_ref_count_zero_list can be used before it is defined?
the thread_local variable is introduced in the next commit, but this commit already uses it?

It doesn't make sense to me. Why not merge the last two commits into single one?

File: compiler-rt/lib/plsan/plsan.cpp [hunk 2]
```diff
@@ -30,46 +30,31 @@ extern "C" void __plsan_store(void **lhs, void *rhs) {
...
 
   if (!is_allocated)
-    return nullptr;
+    return;
 
-  __sanitizer::Vector<void *> *ref_count_zero_addrs =
-      __plsan::free_local_variable(arr_start_addr, size, ret_addr, is_return);
+  __plsan::free_local_variable(arr_start_addr, size, ret_addr, is_return);
 
   // This return will be changed. It have to contain stack trace data.
```
[the remark this answers is not in the document]
Comment (purplepig4657) [reply compiler-rt/lib/plsan/plsan.cpp:42]: Ok

File: compiler-rt/lib/plsan/plsan.cpp (hunk 1 above)
Comment (purplepig4657) [reply compiler-rt/lib/plsan/plsan.cpp:48]: I think you saw commit logs wrong order. local_var_ref_count_zero_list defined last commit.

Comment (hygoni) [reply compiler-rt/lib/plsan/plsan.cpp:45]: Oh, you are right!
My mistake.

Comment (purplepig4657) [reply compiler-rt/lib/plsan/plsan.cpp:45]: And do you think [68f1ff2](https://github.com/hygoni/precise-leak-sanitizer/pull/105/commits/68f1ff2558184936a8de680c8d4a64407283ab46), [a716e12](https://github.com/hygoni/precise-leak-sanitizer/pull/105/commits/a716e12a4ddebde875f8675482fa0d060b08da5c) these 2 commits should be merge?

**Pull request merged**